### DISCUSSION TITLE
How to achieve cross tree communication and data updates ?

### DISCUSSION THREAD

**Discussion by anishmiviewis** [mobxjs/mobx-state-tree#2181] [General] [answered]
I have a setup where I have one global data store and multiple viewModel trees and I want to get a copy of data from global store and use that data in viewmodels and do operation over it but not mutate the global store. How to achieve this kind of behavior. I have created a sand box which has desired intent of mine. I really need help on this. 

Each ViewModels and globalDataRepo is entirely different trees.

Following is the link to the sandbox

[codeSanboxlink](https://codesandbox.io/p/sandbox/mobx-state-tree-playground-forked-nj88wx?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clwzq816q00063b6ji7d9h9df%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clwzq816q00023b6jxm079i1s%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clwzq816q00033b6j25l8zxwj%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clwzq816q00053b6jt6acsbyl%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B72.45098682053636%252C27.549013179463643%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clwzq816q00023b6jxm079i1s%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clwzq816q00013b6j67mnzlzt%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clwzq816q00023b6jxm079i1s%2522%252C%2522activeTabId%2522%253A%2522clwzq816q00013b6j67mnzlzt%2522%257D%252C%2522clwzq816q00053b6jt6acsbyl%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clwzq816q00043b6jcr7klo74%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clwzq816q00053b6jt6acsbyl%2522%252C%2522activeTabId%2522%253A%2522clwzq816q00043b6jcr7klo74%2522%257D%252C%2522clwzq816q00033b6j25l8zxwj%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clwzq816q00033b6j25l8zxwj%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Afalse%252C%2522sidebarPanelSize%2522%253A0%257D)

**Comment by coolsoftwaretyler** [comment, marked as answer]
Hi @anishmiviewis - thanks for the question! Tree nodes cannot be shared across separate trees directly. [Each node can only exist in exactly one tree](https://mobx-state-tree.js.org/concepts/trees#:~:text=However%2C%20each%20node%20can%20only,semantics%20can%20still%20be%20upheld.). 

However, if you want to just share that data across trees, you can always take a snapshot of the items you care about, and create copies in a separate tree. They will be *distinct nodes* from one another, so they won't share state, but this is an easy way to move data around from different trees. Here's an example using `.map(getSnapshot)` based on your CodeSandbox. I also fixed a bug where you updated a task item with an incorrect value (using the number `1` instead of the string `"1"`): https://codesandbox.io/p/sandbox/mobx-state-tree-playground-forked-cmj7yy?file=%2Fsrc%2Findex.js%3A34%2C27

The drawback to this approach is, as I mentioned, you'll end up with copies of your data that will not stay in sync. If that's fine for this use case, because the ViewModel just needs to represent that information visually, then this may be sufficient.

I like using separate view models when they are precisely that: entirely separate view models that can control the behavior of some component or interface. But if you need data sharing between the VMs and your global state, there's no harm in making them part of the same tree, in my opinion.

**Comment by anishmiviewis** [comment]
@coolsoftwaretyler  thank you very much for the solution. Actually in my senario the global store is setup in way to act as an repository outlet which listens to the database changes from the repository layer and gets auto updated when ever data layer changes for that particular Local Database Model and updates  model  of the global store which then is consumed by certain set of views in stack Screens and their respective ViewModels consumes the updated data and displays it or will get the copy of the model from global store  and make changes if the view needs some additional client state for specific scenarios and VM's are never meant to update that particular nature of global store/outlet. So as you suggested I think getSnapShot works perfect for that if I expose it through derivedState ie .views({ taskItems: () => getSnapShot(self.currentJob.taskItems)}) so that VM and Views do not need to worry about calling getSnapShot. And I tried it and works great. Thank you so much for the help :)
